concordium-std/src/state_btree.rs

```diff
@@ -0,0 +1,2111 @@
+            let mut child = self.get_node_mut(*child_id);
+            node = if !child.is_full() {
+                child
+            } else {
+                let larger_child = self.split_child(&mut node, insert_index, &mut child);
+                // Since the child is now split into two, we have to check which one to insert
+                // into.
+                let moved_up_key = &node.keys[insert_index];
+                match moved_up_key.cmp(&key) {
+                    Ordering::Equal => return false,
+                    Ordering::Less => larger_child,
+                    Ordering::Greater => child,
+                }
+            };
+        }
+    }
+
+    /// Internal function for splitting the child node at a given index for a
+    /// given node. This will also mutate the given node adding a new key
+    /// and child after the provided child_index.
+    /// Returns the newly created node.
+    ///
+    /// Assumes:
+    /// - Node is not a leaf and has `child` as child at `child_index`.
+    /// - `child` is at maximum keys (`2 * M - 1`).
+    fn split_child<'b>(
+        &mut self,
+        node: &mut Node<M, K>,
+        child_index: usize,
+        child: &mut Node<M, K>,
+    ) -> StateRefMut<'b, Node<M, K>, StateApi>
+    where
+        K: Serialize + Ord, {
+        let split_index = Node::<M, K>::MINIMUM_KEY_LEN + 1;
+        let (new_larger_sibling_id, new_larger_sibling) = self.create_node(
+            child.keys.split_off(split_index),
+            if child.is_leaf() {
+                Vec::new()
+            } else {
+                child.children.split_off(split_index)
+            },
+        );
+        let key = child.keys.pop().unwrap_abort();
+        node.children.insert(child_index + 1, new_larger_sibling_id);
+        node.keys.insert(child_index, key);
+        new_larger_sibling
+    }
+
+    /// Internal function for looking up a node in the tree.
+    /// This assumes the node is present and traps if this is not the case.
+    fn get_node<Key>(&self, node_id: NodeId) -> Node<M, Key>
+    where
+        Key: Deserial, {
+        let key = node_id.as_key(&self.prefix);
+        let mut entry = self.state_api.lookup_entry(&key).unwrap_abort();
+        entry.get().unwrap_abort()
+    }
+
+    /// Internal function for looking up a node, providing mutable access.
+    /// This assumes the node is present and traps if this is not the case.
+    fn get_node_mut<'b>(&mut self, node_id: NodeId) -> StateRefMut<'b, Node<M, K>, StateApi>
+    where
+        K: Serial, {
+        let key = node_id.as_key(&self.prefix);
+        let entry = self.state_api.lookup_entry(&key).unwrap_abort();
+        StateRefMut::new(entry, self.state_api.clone())
+    }
+
+    /// Construct a string for displaying the btree and debug information.
+    /// Should only be used while debugging and testing the btree itself.
+    #[cfg(feature = "internal-wasm-test")]
+    pub(crate) fn debug(&self) -> String
+    where
```

Comment:
Since this is printing the internal structure of the btree, I wanted to avoid implementing `Debug`.
Don't think the user should ever need this specific format.

Reply:
It would only be implemented under this feature, at least that was my suggestion.